sdks/java/core/src/main/java/org/apache/beam/sdk/metrics/MetricsContainers.java

```diff
@@ -0,0 +1,55 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.sdk.metrics;
+
+import java.io.Serializable;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+/**
+ * Metrics containers by step.
+ */
+public class MetricsContainers implements Serializable {
+  private Map<String, MetricsContainer> metricsContainers;
+
+  public MetricsContainers() {
+    this.metricsContainers = new ConcurrentHashMap<>();
+  }
+
+  public MetricsContainer getContainer(String stepName) {
+    if (!metricsContainers.containsKey(stepName)) {
+      metricsContainers.put(stepName, new MetricsContainer(stepName));
+    }
+    return metricsContainers.get(stepName);
+  }
+
+  public void updateAll(MetricsContainers other) {
+    for (Map.Entry<String, MetricsContainer> container : other.metricsContainers.entrySet()) {
```

Comment:
Does this need to do any locking? Either way, should extend the documentation to describe how this should be used by the runner, and what requirements/guarantees it provides in terms of thread safety.